src/analysis/convex/cone/proper.lean

```diff
@@ -46,8 +47,76 @@ protected def closure (K : convex_cone ℝ E) : convex_cone ℝ E :=
 
 @[simp, norm_cast] lemma coe_closure (K : convex_cone ℝ E) : (K.closure : set E) = closure K := rfl
 
-protected lemma mem_closure {K : convex_cone ℝ E} {a : E} :
+@[simp] protected lemma mem_closure {K : convex_cone ℝ E} {a : E} :
   a ∈ K.closure ↔ a ∈ closure (K : set E) :=
 iff.rfl
 
 end convex_cone
+
+section definitions
+
+/-- A proper cone is a convex cone `K` that is nonempty and closed. Proper cones have the nice
+property that the dual of the dual of a proper cone is itself. This makes them useful for defining
+cone programs and proving duality theorems. -/
+structure proper_cone (E : Type*) [normed_add_comm_group E] [inner_product_space ℝ E]
+  [complete_space E] extends convex_cone ℝ E :=
+(nonempty'  : (carrier : set E).nonempty)
+(is_closed' : is_closed (carrier : set E))
+
+end definitions
+
+namespace proper_cone
+
+section complete_space
+
+variables {E : Type*} [normed_add_comm_group E] [inner_product_space ℝ E] [complete_space E]
+variables {F : Type*} [normed_add_comm_group F] [inner_product_space ℝ F] [complete_space F]
+
+instance : has_coe (proper_cone E) (convex_cone ℝ E) := ⟨λ K, K.1⟩
+
+instance : has_mem E (proper_cone E) := ⟨λ e K, e ∈ (K : convex_cone ℝ E) ⟩
+
+noncomputable instance : has_zero (proper_cone E) := ⟨ ⟨0, ⟨0, rfl⟩, is_closed_singleton⟩ ⟩
+
+noncomputable instance : inhabited (proper_cone E) := ⟨0⟩
+
+@[simp] lemma mem_coe {x : E} {K : proper_cone E} : x ∈ K ↔ x ∈ (K : convex_cone ℝ E) := iff.rfl
+
+lemma nonempty (K : proper_cone E) : (K : set E).nonempty := K.nonempty'
+
+lemma is_closed (K : proper_cone E) : is_closed (K : set E) := K.is_closed'
+
+lemma pointed (K : proper_cone E) : (K : convex_cone ℝ E).pointed :=
+(K : convex_cone ℝ E).pointed_of_nonempty_of_is_closed K.nonempty K.is_closed
+
+@[ext] lemma ext {S T : proper_cone E} (h : (S : convex_cone ℝ E) = T) : S = T :=
+by cases S; cases T; congr'
+
+/-- We define `K*` to be the inner dual cone of `K`. -/
+instance : has_star (proper_cone E) := ⟨ λ K,
+  ⟨ (K : set E).inner_dual_cone, ⟨0, pointed_inner_dual_cone _⟩, is_closed_inner_dual_cone _ ⟩ ⟩
+
+/-- The dual of the dual of a proper cone is itself. -/
+instance : has_involutive_star (proper_cone E) :=
+{ star := has_star.star,
+  star_involutive := λ K, proper_cone.ext $
+    (K : convex_cone ℝ E).inner_dual_cone_of_inner_dual_cone_eq_self K.nonempty K.is_closed }
+
+@[simp] lemma coe_star (K : proper_cone E) : ↑(star K) = (K : set E).inner_dual_cone := rfl
+
+@[simp] lemma mem_star {K : proper_cone E} {y : E} :
+  y ∈ star K ↔ ∀ x ∈ (K : set E), 0 ≤ ⟪x, y⟫_ℝ :=
+by rw [mem_coe, coe_star, mem_inner_dual_cone _ _]
+
+/-- The closure of image of a proper cone under a continuous `ℝ`-linear map is a proper cone. -/
+noncomputable def map (f : E →L[ℝ] F) (K : proper_cone E) : proper_cone F :=
```

Comment:
Can this be bundled in any way? Maybe it can be
```suggestion
noncomputable def map (f : E →L[ℝ] F) : proper_cone E →L[ℝ] proper_cone F :=
```
once you've defined pointwise operations on (proper) cones?

Reply:
I see. This should probably also be done for `convex_cone.map`. Is it ok if I leave a TODO for now and bundle this and the `convex_cone.map` in a separate PR?

Reply:
Am happy with a TODO, yes.